app/merginapi.cpp

```diff
@@ -5,29 +5,136 @@
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: listProjects" );
+        return;
+    }
+
     QNetworkRequest request;
     // projects filtered by tag "input_use"
     QUrl url(mApiRoot + "/v1/project?tags=input_use");
-
     request.setUrl(url);
     request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
 
     QNetworkReply *reply = mManager.get(request);
     connect(reply, &QNetworkReply::finished, this, &MerginApi::listProjectsReplyFinished);
 }
 
+void MerginApi::downloadProject(QString projectName)
+{
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: downloadProject" );
+    }
+
+    QNetworkRequest request;
+    QUrl url(mApiRoot + "/v1/project/download/" + projectName);
+
+    if (mPendingRequests.contains(url)) {
+        QString errorMsg = QStringLiteral("Download request for %1 is already pending.").arg(projectName);
+        qDebug() << errorMsg;
+        emit networkErrorOccurred( errorMsg, "Mergin API error: downloadProject" );
+        return;
+    }
+
+    request.setUrl(url);
+    request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
+
+    QNetworkReply *reply = mManager.get(request);
+    mPendingRequests.insert(url, projectName);
+    connect(reply, &QNetworkReply::finished, this, &MerginApi::downloadProjectReplyFinished);
+}
+
+void MerginApi::updateProject(QString projectName)
+{
+
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: projectInfo" );
+    }
+
+    QNetworkRequest request;
+    QUrl url(mApiRoot + "/v1/project/" + projectName);
+
+    request.setUrl(url);
+    request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
+
+    QNetworkReply *reply = mManager.get(request);
+    connect(reply, &QNetworkReply::finished, this, &MerginApi::updateInfoReplyFinished);
+
+}
+
+void MerginApi::downloadProjectFiles(QString projectName, QByteArray json)
+{
+    if (mToken.isEmpty()) {
+        emit networkErrorOccurred( "Auth token is invalid", "Mergin API error: fetchProject" );
+        return;
+    }
+
+    QNetworkRequest request;
+    QUrl url(mApiRoot + "/v1/project/fetch/" + projectName);
+    request.setUrl(url);
+    request.setRawHeader("Authorization", QByteArray("Basic " + mToken));
+    request.setRawHeader("Content-Type", "application/json");
+    request.setRawHeader("Accept", "application/json");
+    mPendingRequests.insert(url, projectName);
+
+    QNetworkReply *reply = mManager.post(request, json);
+    connect(reply, &QNetworkReply::finished, this, &MerginApi::downloadProjectReplyFinished);
+}
+
+ProjectList MerginApi::updateMerginProjectList(ProjectList serverProjects)
+{
+    QHash<QString, QString> projectUpdates;
+    for (std::shared_ptr<MerginProject> project: mMerginProjects) {
+        projectUpdates.insert(project.get()->name, project.get()->updated);
```

Comment:
use `project->name` instead of `project.get()->name`  ... shared_ptr also overrides `->` operator so it "feels" like working with a raw pointer.

same thing in for-cycle below and in setUpdateToProject